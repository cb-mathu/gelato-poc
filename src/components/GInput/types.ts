import { ExtractPropTypes, HTMLAttributes } from 'vue';
import { MergedCoreInputProps, InputTypes } from '../core/Input/types';
import { INPUT_PROPS } from './constants';

export { InputTypes };

export type InputProps = ExtractPropTypes<typeof INPUT_PROPS>;
export type InputErrorProps = ExtractPropTypes<Pick<InputProps, 'errorMessage' | 'id'>> & HTMLAttributes;
export type MergedInputProps = Partial<MergedCoreInputProps & InputProps>;
