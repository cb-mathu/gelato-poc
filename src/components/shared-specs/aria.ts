/* istanbul ignore file */
import { render } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';

export const FormFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Form Field ARIA`, () => {
    test('renders aria-labelledBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAttribute('aria-labelledBy');
    });

    test('renders aria-invalid', async () => {
      render(Component({ invalid: true }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toBeInvalid();
    });

    test('renders aria-errorMessage', async () => {
      const errorMessage = 'This is an error message, ooh poor you!';
      render(Component({ invalid: true, errorMessage: errorMessage }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveErrorMessage(errorMessage);
    });
  });

export const ControlFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Control Field ARIA`, () => {
    test('renders aria-checked', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByRole(role, { checked: true })).toBeTruthy();
    });
    FormFieldAria(Component, label, role);
  });

export const InputFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Input Field ARIA`, () => {
    test('renders aria-describedBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAccessibleDescription();
    });
    FormFieldAria(Component, label, role);
  });
