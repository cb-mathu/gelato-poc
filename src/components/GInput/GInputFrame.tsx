import { defineComponent } from 'vue';
import { INPUT_PROPS } from './constants';
import { GNotification, NotificationVariant, NotificationSize } from '../GNotification';
import { InputErrorProps } from './types';
import { useFormAria, useFormAriaIds } from '../../utils/accessibility';
import { useId, useInput } from '../../composables';
import { isNotEmpty } from '../../utils/common';

export const GInputFrameErrorMessage = (props: InputErrorProps) => {
  return (
    <div id={props.id}>
      <GNotification size={NotificationSize.Small} variant={NotificationVariant.Alert}>
        <span>{props.errorMessage}</span>
      </GNotification>
    </div>
  );
};

export default defineComponent({
  name: 'GInputFrame',
  props: INPUT_PROPS,
  emits: ['blur'],
  inheritAttrs: false,

  setup(props, { attrs, slots }) {
    const { labelId, descriptionId, errorId } = useFormAriaIds(props.id);

    const { internalValue, ...listeners } = useInput({
      type: props.type,
      props,
    });

    const propsWeControl: any = {
      id: props.id || useId('GInputFrame'),
    };

    return () => {
      const label = props.label || slots.label?.();
      const description = props.description || slots.default?.();
      const attributes = { ...attrs, ...listeners, ...props, ...propsWeControl };
      const classes = {
        [`g-input`]: true,
        [`g-input--invalid`]: props.invalid,
        [`g-input--disabled`]: props.disabled,
      };
      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          describedBy: descriptionId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
      };

      return (
        <div class={classes}>
          {isNotEmpty(label) && (
            <label class="g-input__label" id={labelId} for={attributes.id}>
              {label}
            </label>
          )}
          {isNotEmpty(description) && (
            <div class="g-input__desc" id={descriptionId}>
              {description}
            </div>
          )}
          <div class="g-input__group">
            {isNotEmpty(slots.prefix?.()) && <span class="g-input__prefix">{slots.prefix?.()}</span>}
            {isNotEmpty(slots.suffix?.()) && <span class="g-input__suffix">{slots.suffix?.()}</span>}
            <div {...attributes} {...ariaAttributes} class="g-input__element" />
          </div>
          {props.invalid && props.errorMessage && <GInputFrameErrorMessage id={errorId} errorMessage={props.errorMessage} />}
        </div>
      );
    };
  },
});
