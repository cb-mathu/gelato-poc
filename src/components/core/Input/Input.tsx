import { defineComponent, watchEffect } from 'vue';
import { INPUT_NAME, CORE_INPUT_PROPS } from './constants';
import { useId, useInput } from '../../../composables';
import { MergedCoreInputProps } from './types';

const Input = defineComponent({
  name: INPUT_NAME,
  props: CORE_INPUT_PROPS,
  emits: ['update:modelValue', 'blur'],

  setup(props, { attrs }) {
    const { internalValue, ...listeners } = useInput({ type: props.type, props });

    return () => {
      const propsWeControl: any = {
        id: props.id || useId(INPUT_NAME),
      };
      const attributes = { ...attrs, ...listeners, ...props, ...propsWeControl };
      return <input {...attributes} value={internalValue.value} />;
    };
  },
});

export default Input as {
  new (): {
    $props: MergedCoreInputProps;
  };
};
