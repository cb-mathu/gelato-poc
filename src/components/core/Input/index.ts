import { InputTypes } from './types';
import Input from './Input';
export * from './constants';
export { InputTypes, Input };
