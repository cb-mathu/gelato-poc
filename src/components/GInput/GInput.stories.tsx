import { ref } from 'vue';
import GInput from './GInput';
import { InputTypes } from './types';
import { GNotification } from '../GNotification';
import { MergedInputProps } from './types';
import { Input } from '../core/Input';

export default {
  title: 'Component/Forms/GInput',
  component: GInput,
  argTypes: {
    name: {
      default: 'Name of the field',
    },
    label: {
      description: 'Label for the field',
      control: {
        type: 'text',
      },
    },
    description: {
      description: 'Description for the field',
      control: {
        type: 'text',
      },
    },
    type: {
      description: 'Input types',
      options: InputTypes,
      defaultValue: InputTypes.Text,
      control: {
        type: 'select',
      },
    },
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
    autofocus: {
      description: 'Autofocus the field',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    autocomplete: {
      description: 'Autcomplete for the field',
      defaultValue: 'off',
      control: {
        type: 'text',
      },
    },
    required: {
      description: 'Field is required',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      description: 'Field is readonly',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    modelValue: {
      description: 'v-model value',
      control: {
        type: 'text',
      },
    },
    id: {
      description: 'Id of the field',
      control: {
        type: 'text',
      },
    },
    placeholder: {
      description: 'placeholder of the field',
      control: {
        type: 'text',
      },
    },
    class: {
      description: 'Class of the field',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    invalid: {
      description: 'Field is invalid',
      defaulValue: false,
      control: {
        type: 'boolean',
      },
    },
    errorMessage: {
      description: 'Error message for invalid field (Applicable only if invalid is set)',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    min: {
      description: 'Min value (Applicable for Number field)',
      defaulValue: -Infinity,
      control: {
        type: 'number',
      },
    },
    max: {
      description: 'Max value (Applicable for Number field)',
      defaulValue: Infinity,
      control: {
        type: 'number',
      },
    },
    maxlength: {
      description: 'Min value (Applicable for Text field)',
      control: {
        type: 'number',
      },
    },
  },
};

const Template = (args: any) => ({
  setup() {
    let input = ref(args.defaultValue || 'John doe');

    const attributes = {
      ...args,
      disabled: args.disabled,
      readonly: args.readonly,
      invalid: args.invalid,
      errorMessage: args.errorMessage,
    };

    const formatter = (val: string) => val.toUpperCase();

    return () => (
      <div>
        <div class="mb-4">
          <GInput {...attributes} v-model={input.value} id="input" label={args.label} description={args.description}>
            {{
              label: () => args.labelSlot,
              default: () => args.descriptionSlot,
              prefix: () => args.prefixSlot && <args.prefixSlot />,
              suffix: () => args.suffixSlot && <args.suffixSlot />,
            }}
          </GInput>
        </div>
        <div class="mb-4">
          <div>
            <label>GInput with lazy and formatter</label>
          </div>
          <GInput v-model={[input.value, ['lazy']]} formatter={formatter} />
        </div>
        <div class="mb-4">
          <div>
            <label>Core Input</label>
          </div>
          <Input v-model={input.value} />
        </div>
        <div class="mb-4">
          <div>
            <label>Core Input Lazy</label>
          </div>
          <Input v-model={[input.value, ['lazy']]} />
        </div>
        <div class="mb-4">
          <div>
            <label>Core Input Formatter</label>
          </div>
          <Input v-model={input.value} formatter={formatter} />
        </div>
        <GNotification>
          Entered Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});

const testLabel = 'First Name';
const testDescription =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin';

//Story
export const Story = Template.bind({});

export const DefaultInput = (props?: MergedInputProps, slots?: any) => (
  <GInput key="default" modelValue="John" {...(props as any)} v-slots={slots} />
);

//Number
export const NumberInput = () => <GInput type={InputTypes.Number} modelValue={10} />;

// With Label
export const InputWithLabel = (props?: any) => <GInput modelValue="John" label={testLabel} {...props} />;

// With description
export const InputWithDescription = () => <GInput label={testLabel} modelValue="John" description={testDescription} />;

// with Slots
export const InputWithSlots: any = (props?: any) => (
  <GInput modelValue="John" {...props} v-slots={{ label: () => testLabel, default: () => testDescription }} />
);

export const addonImage = () => (
  <svg xmlns="http://www.w3.org/2000/svg" width="20px" viewBox="0 0 20 20" fill="currentColor">
    <path
      fill-rule="evenodd"
      d="M5 9V7a5 5 0 0110 0v2a2 2 0 012 2v5a2 2 0 01-2 2H5a2 2 0 01-2-2v-5a2 2 0 012-2zm8-2v2H7V7a3 3 0 016 0z"
      clip-rule="evenodd"
    ></path>
  </svg>
);

// With Prefix
export const InputWithPrefix = () => <GInput label={testLabel} description={testDescription} v-slots={{ prefix: addonImage }} />;

// With Suffix
export const InputWithSuffix = () => <GInput label={testLabel} description={testDescription} v-slots={{ suffix: addonImage }} />;

// With Prefix and Suffix
export const InputWithPrefixAndSuffix = () => (
  <GInput label={testLabel} description={testDescription} v-slots={{ prefix: addonImage, suffix: addonImage }} />
);

// Disabled
export const DisabledInput: any = () => <GInput modelValue="John" label={testLabel} disabled />;

// Readonly
export const ReadonlyInput: any = () => <GInput modelValue="John" label={testLabel} readonly />;

// Invalid
export const InvalidInput: any = () => (
  <GInput modelValue="John" label={testLabel} invalid errorMessage="This is an error message, ooh poor you!" />
);

// Lazy
export const LazyInput = () => ({
  setup() {
    let input = ref('John doe');

    return () => (
      <div>
        <div style="margin-bottom: 20px;">
          <GInput label={testLabel} v-model={[input.value, ['lazy']]} />
        </div>
        <GNotification>
          Entered Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});
