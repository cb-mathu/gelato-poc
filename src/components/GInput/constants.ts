import { PropType } from 'vue';
import { CORE_INPUT_PROPS } from '../core/Input';

export const INPUT_PROPS = {
  ...CORE_INPUT_PROPS,
  label: {
    type: String,
  },
  description: {
    type: String,
  },
  'onUpdate:modelValue': Function as PropType<(value: Event) => void>,
  onInput: Function as PropType<(event: Event) => void>,
  onKeydown: Function as PropType<(event: KeyboardEvent) => void>,
};
