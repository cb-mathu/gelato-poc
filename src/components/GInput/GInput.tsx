import { defineComponent } from 'vue';
import { Input } from '../core/Input';
import { INPUT_PROPS } from './constants';
import { GNotification, NotificationVariant, NotificationSize } from '../GNotification';
import { InputErrorProps } from './types';
import { useFormAria, useFormAriaIds } from '../../utils/accessibility';
import { useId, useInput } from '../../composables';
import { isNotEmpty } from '../../utils/common';

export const GInputErrorMessage = (props: InputErrorProps) => {
  return (
    <div id={props.id}>
      <GNotification size={NotificationSize.Small} variant={NotificationVariant.Alert}>
        <span>{props.errorMessage}</span>
      </GNotification>
    </div>
  );
};

export default defineComponent({
  name: 'GInput',
  props: INPUT_PROPS,
  emits: ['update:modelValue', 'blur'],
  inheritAttrs: false,

  setup(props, { attrs, slots }) {
    const { labelId, descriptionId, errorId } = useFormAriaIds(props.id);

    const { internalValue, ...listeners } = useInput({
      type: props.type,
      props,
    });

    return () => {
      const label = props.label || slots.label?.();
      const description = props.description || slots.default?.();
      const propsWeControl: any = {
        id: props.id || useId('GInput'),
      };
      const attributes = { ...attrs, ...listeners, ...props, ...propsWeControl };
      const classes = {
        [`g-input`]: true,
        [`g-input--invalid`]: props.invalid,
        [`g-input--disabled`]: props.disabled,
      };
      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          describedBy: descriptionId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
      };

      return (
        <div class={classes}>
          {isNotEmpty(label) && (
            <label class="g-input__label" id={labelId} for={attributes.id}>
              {label}
            </label>
          )}
          {isNotEmpty(description) && (
            <div class="g-input__desc" id={descriptionId}>
              {description}
            </div>
          )}
          <div class="g-input__group">
            {isNotEmpty(slots.prefix?.()) && <span class="g-input__prefix">{slots.prefix?.()}</span>}
            {isNotEmpty(slots.suffix?.()) && <span class="g-input__suffix">{slots.suffix?.()}</span>}
            <Input {...attributes} {...ariaAttributes} modelValue={internalValue.value} class="g-input__element" />
          </div>
          {props.invalid && props.errorMessage && <GInputErrorMessage id={errorId} errorMessage={props.errorMessage} />}
        </div>
      );
    };
  },
});
