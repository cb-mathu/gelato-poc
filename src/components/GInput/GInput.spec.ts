import { render } from '@testing-library/vue';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import * as stories from './GInput.stories';
import { InputFieldAria } from '../shared-specs/aria';
import { InputTypes } from '../core/Input';
const {
  DefaultInput,
  NumberInput,
  InputWithLabel,
  InputWithDescription,
  InputWithSlots,
  InputWithPrefix,
  InputWithSuffix,
  DisabledInput,
  ReadonlyInput,
  InvalidInput,
  LazyInput,
} = stories;

const testLabel = 'First Name';
const testErroMessage = 'This is an error message, ooh poor you!';

describe('GInput: ', () => {
  test('renders with default args', () => {
    const { container } = render(DefaultInput());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders Number field', () => {
    const { container } = render(NumberInput());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders description', () => {
    const { container } = render(InputWithDescription());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders slots', () => {
    const { container } = render(InputWithSlots());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders hidden field', () => {
    render(DefaultInput({ type: InputTypes.Hidden, label: testLabel }));
    expect(screen.getByLabelText(testLabel)).not.toBeVisible();
  });

  test('renders in disabled state', () => {
    render(DisabledInput());
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('disabled');
  });

  test('renders in readonly state', () => {
    render(ReadonlyInput());
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('readonly');
  });

  test('renders in invalid state', () => {
    render(InvalidInput());
    expect(screen.getByText(testErroMessage)).toBeVisible();
  });

  test('pressing tab will focus input', () => {
    render(InputWithLabel());
    userEvent.tab();
    expect(screen.getByLabelText(testLabel)).toHaveFocus();
  });

  test('focuses input when label is clicked', () => {
    render(InputWithLabel());
    const element = screen.getByText(testLabel);
    userEvent.click(element);
    expect(screen.getByLabelText(testLabel)).toHaveFocus();
  });

  test('lazy: updates input on change event', () => {
    const { container } = render(LazyInput());
    const inputValue = ' Alex';
    userEvent.type(screen.getByLabelText(testLabel), inputValue);
    container.parentElement?.click();
    expect(screen.getByLabelText(testLabel)).toHaveValue('John doe Alex');
  });
});

describe('GInput Group: ', () => {
  test('renders prefix', async () => {
    render(InputWithPrefix());
    expect(screen.getByLabelText(testLabel).parentElement?.firstChild).not.toBeEmptyDOMElement();
  });

  test('renders suffix', async () => {
    render(InputWithSuffix());
    expect(screen.getByLabelText(testLabel).previousSibling).not.toBeEmptyDOMElement();
  });
});

InputFieldAria(InputWithSlots, testLabel, 'input');
