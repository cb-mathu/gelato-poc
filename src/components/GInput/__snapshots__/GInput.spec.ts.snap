// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GInput:  renders Number field 1`] = `
<div
  class="g-input"
>
  <!---->
  <!---->
  <div
    class="g-input__group"
  >
    <!---->
    <!---->
    <input
      aria-describedby="description_6"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_5"
      autocomplete="off"
      class="g-input__element"
      id="GInput_8"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="10"
      name=""
      placeholder=""
      type="number"
    />
  </div>
  <!---->
</div>
`;

exports[`GInput:  renders description 1`] = `
<div
  class="g-input"
>
  <label
    class="g-input__label"
    for="GInput_12"
    id="label_9"
  >
    First Name
  </label>
  <div
    class="g-input__desc"
    id="description_10"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin
  </div>
  <div
    class="g-input__group"
  >
    <!---->
    <!---->
    <input
      aria-describedby="description_10"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_9"
      autocomplete="off"
      class="g-input__element"
      description="Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin"
      id="GInput_12"
      invalid="false"
      label="First Name"
      modelmodifiers="[object Object]"
      modelvalue="John"
      name=""
      placeholder=""
      type="text"
    />
  </div>
  <!---->
</div>
`;

exports[`GInput:  renders slots 1`] = `
<div
  class="g-input"
>
  <label
    class="g-input__label"
    for="GInput_16"
    id="label_13"
  >
    
    First Name
    
  </label>
  <div
    class="g-input__desc"
    id="description_14"
  >
    
    Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin
    
  </div>
  <div
    class="g-input__group"
  >
    <!---->
    <!---->
    <input
      aria-describedby="description_14"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_13"
      autocomplete="off"
      class="g-input__element"
      id="GInput_16"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="John"
      name=""
      placeholder=""
      type="text"
    />
  </div>
  <!---->
</div>
`;

exports[`GInput:  renders with default args 1`] = `
<div
  class="g-input"
>
  <!---->
  <!---->
  <div
    class="g-input__group"
  >
    <!---->
    <!---->
    <input
      aria-describedby="description_2"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_1"
      autocomplete="off"
      class="g-input__element"
      id="GInput_4"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="John"
      name=""
      placeholder=""
      type="text"
    />
  </div>
  <!---->
</div>
`;
